design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+particular kind of resource claim, including specifying how resource claim
+fields should relate to the dynamically provisioned managed resources.
+
+### Example: Azure MySQL Server with Virtual Network Rule
+
+The below example uses a user defined resource claim (`MySQLInstance`) along
+with a `ResourceConfigurationClass` to produce a `ResourceConfiguration`
+consisting of an Azure MySQL server and a virtual network rule.
+
+```yaml
+---
+# Note that this MySQLInstance is in the database.example.org API group (not
+# database.crossplane.io). It includes fields not supported by the contemporary
+# MySQLInstance.database.crossplane.io claim kind.
+apiVersion: database.example.org/v1alpha1
+kind: MySQLInstance
+metadata:
+  namespace: default
+  name: sql
+  annotations:
+    # The external name annotation is copied to all managed resources that are
+    # dynamically provisioned by this claim.
+    crossplane.io/external-name: very-private-database
+spec:
+  engineVersion: "5.7"
+  storageGB: 10
+  region: us-west
+  # Resource claim custom resources must support Crossplane's claim binding
+  # metadata; they must have a writeConnectionSecretToRef, classRef,
+  # resourceRef, bindingPhase, and status conditions.
+  writeConnectionSecretToRef:
+    name: sql
+  classRef:
+    apiVersion: infrastructure.crossplane.io/v1alpha1
+    kind: ResourceConfigurationClass
+    name: private-mysql-server
+  resourceRef:
+    apiVersion: infrastructure.crossplane.io/v1alpha1
+    kind: ResourceConfiguration
+    name: default-sql-gdm0w
+---
+apiVersion: infrastructure.crossplane.io/v1alpha1
+kind: ResourceConfigurationClass
+metadata:
+  name: private-mysql-server
+spec:
+  # Each ResourceConfigurationClass defines how resources are dynamically
+  # provisioned for exactly one kind of resource claim; there is a one-to-one
+  # claim-kind-to-configuration-class-instance relationship.
+  claim:
+    apiVersion: database.example.org/v1alpha1
+    kind: MySQLInstance
+  # The array of primitive managed resources that will be dynamically
+  # provisioned to satisfy MySQLInstance claims that use this class. Each entry
+  # is an object containing a 'base' and an optional 'patch'. The base must be a
+  # valid managed resource (i.e. a defined custom resource that supports
+  # Crossplane's claim binding metadata). The patch specifies how the claim's
+  # fields may be used to extend or override the base's fields.
+  resources:
+  - base:
+      apiVersion: azure.crossplane.io/v1alpha3
+      kind: ResourceGroup
+      metadata:
+      spec:
+        # Required base fields (such as location) must specify a default value,
+        # even if it may be overridden by a patch.
+        location: West US
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    # Each patch entry relates a claim field to a resource field, by fieldRef.
+    patch:
+    - fromClaimField: ".spec.region"
+      toResourceField: ".spec.forProvider.location"
```

Comment:
`fromClaim` and `toResource` could read better.

Reply:
One one hand I like the succinctness of `fromClaim` and `toResource`, but on the other hand I worry about losing the context that it's a field ref.

I also considered something more verbose (but more flexible in line with API conventions?):

```yaml
- fromClaim:
     fieldRef: ".spec.region"
    toResource:
      fieldRef: ".spec.forProvider.location"
```

Reply:
`metav1.ObjectReference` has a field called `fieldPath` for this. So, maybe this would be more inline:
```yaml
- fromClaim:
    fieldPath: ".spec.region"
  toResource:
    fieldPath: ".spec.forProvider.location"
```
I don't have a strong opinion though.

Reply:
Ah - yes `fieldPath` was what I was thinking of. Thanks!

Reply:
More spitballing - another shorter version with less context could be:

```yaml
fromFieldPath: ".spec.region"
toFieldPath: ".spec.forProvider.location"
```

Reply:
I feel like `fromFieldPath` and `toFieldPath` are a reasonable here. They're more succinct than the downward API equivalents, and they're generic enough to handle the case in which "from" could be an application or infrastructure CR.